parser/src/main/scala/org/bykn/bosatsu/parser/Parser.scala

```diff
@@ -0,0 +1,938 @@
+  }
+
+  def oneOf[A](ps: List[Parser[A]]): Parser[A] = {
+    @annotation.tailrec
+    def flatten(ls: List[Parser[A]], acc: List[Parser[A]]): Parser[A] =
+      ls match {
+        case Nil =>
+          acc match {
+            case h :: Nil => h
+            case other => Impl.OneOf(other.reverse)
+          }
+        case Impl.OneOf(ps) :: rest =>
+          flatten(ps ::: rest, acc)
+        case notOneOf :: rest =>
+          flatten(rest, notOneOf :: acc)
+      }
+
+    flatten(ps, Nil)
+  }
+
+  /**
+   * Parse the next len characters where len > 0
+   */
+  def length(len: Int): Parser1[String] =
+    Impl.Length(len)
+
+  /**
+   * Repeat this parser 0 or more times
+   * note: this can wind up parsing nothing
+   */
+  def rep[A](p1: Parser1[A]): Parser[List[A]] =
+    Impl.Rep(p1)
+
+  /**
+   * Repeat this parser 1 or more times
+   */
+  def rep1[A](p1: Parser1[A], min: Int): Parser1[NonEmptyList[A]] =
+    Impl.Rep1(p1, min)
+
+  def product10[A, B](first: Parser1[A], second: Parser[B]): Parser1[(A, B)] =
+    Impl.Prod1(first, second)
+
+  def product01[A, B](first: Parser[A], second: Parser1[B]): Parser1[(A, B)] =
+    Impl.Prod1(first, second)
+
+  def product[A, B](first: Parser[A], second: Parser[B]): Parser[(A, B)] =
+    Impl.Prod(first, second)
+
+  def map[A, B](p: Parser[A])(fn: A => B): Parser[B] =
+    Impl.Map(p, fn)
+
+  def map1[A, B](p: Parser1[A])(fn: A => B): Parser1[B] =
+    Impl.Map1(p, fn)
+
+  def flatMap[A, B](pa: Parser[A])(fn: A => Parser[B]): Parser[B] =
+    Impl.FlatMap(pa, fn)
+
+  def flatMap10[A, B](pa: Parser1[A])(fn: A => Parser[B]): Parser1[B] =
+    Impl.FlatMap1(pa, fn)
+
+  def flatMap01[A, B](pa: Parser[A])(fn: A => Parser1[B]): Parser1[B] =
+    Impl.FlatMap1(pa, fn)
+
+  def tailRecM[A, B](init: A)(fn: A => Parser[Either[A, B]]): Parser[B] =
+    Impl.TailRecM(init, fn)
+
+  def tailRecM1[A, B](init: A)(fn: A => Parser1[Either[A, B]]): Parser1[B] =
+    Impl.TailRecM1(init, fn)
+
+  def defer1[A](pa: => Parser1[A]): Parser1[A] =
+    Impl.Defer1(() => pa)
+
+  def defer[A](pa: => Parser[A]): Parser[A] =
+    Impl.Defer(() => pa)
+
+  val Fail: Parser1[Nothing] = oneOf1(Nil)
+
+  def fail[A]: Parser1[A] = Fail
+
+  def charIn(c0: Char, cs: Char*): Parser1[Char] =
+    charIn1(c0 :: cs.toList)
+
+  def charIn1(cs: Iterable[Char]): Parser1[Char] = {
+    val ary = cs.toArray
+    java.util.Arrays.sort(ary)
+    Impl.CharIn(ary(0).toInt, BitSetUtil.bitSetFor(ary), Impl.rangesFor(ary))
+  }
+
+  def charIn(cs: Iterable[Char]): Parser[Char] =
+    if (cs.isEmpty) fail
+    else charIn1(cs)
+
+  def charWhere(fn: Char => Boolean): Parser[Char] = {
+    // we use defer to avoid enumerating this function unless needed
+    defer(charIn(Impl.allChars.filter(fn)))
+  }
+
+  def charWhere1(fn: Char => Boolean): Parser1[Char] =
+    // we use defer to avoid enumerating this function unless needed
+    defer1(charIn1(Impl.allChars.filter(fn)))
+
+  def void[A](pa: Parser[A]): Parser[Unit] =
```

Comment:
Add a comment that void is a considerable optimization as is string on a large parse since it removes the internal allocations to build up structure.